src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -140,107 +139,77 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
-
-                        await pipe(pushContext).ConfigureAwait(false);
-
-                        rejectMessage = tokenSource.IsCancellationRequested;
-                    }
-                }
-
-                if (rejectMessage)
-                {
-                    try
-                    {
-                        await RejectMessage(channel, message.DeliveryTag).ConfigureAwait(false);
-                    }
-                    catch (AlreadyClosedException ex)
-                    {
-                        Logger.Warn($"Attempt to reject message {messageId} failed because the channel was closed. The message will be requeued.", ex);
-                    }
-                }
-                else
+                using (var tokenSource = new CancellationTokenSource())
                 {
-                    try
+                    var pushContext = new PushContext(messageId, headers, new MemoryStream(message.Body ?? new byte[0]), new TransportTransaction(), tokenSource, new ContextBag());
+                    await pipe(pushContext).ConfigureAwait(false);
+                    if (tokenSource.IsCancellationRequested)
                     {
-                        await AcknowledgeMessage(channel, message.DeliveryTag).ConfigureAwait(false);
+                        await TryRejectMessage(message.DeliveryTag, messageId).ConfigureAwait(false);
                     }
-                    catch (AlreadyClosedException ex)
+                    else
                     {
-                        Logger.Warn($"Attempt to acknowledge message {messageId} failed because the channel was closed. The message will be requeued.", ex);
+                        await TryAcknowledgeMessage(message.DeliveryTag, messageId).ConfigureAwait(false);
                     }
                 }
             }
             catch (Exception ex)
             {
                 Logger.Warn($"Error while attempting to process message {messageId}. The message will be rejected.", ex);
-
-                try
-                {
-                    await RejectMessage(channel, message.DeliveryTag).ConfigureAwait(false);
-                }
-                catch (AlreadyClosedException ex2)
-                {
-                    Logger.Warn($"Attempt to reject message {messageId} failed because the channel was closed. The message will be requeued.", ex2);
-                }
+                await TryRejectMessage(message.DeliveryTag, messageId).ConfigureAwait(false);
             }
-            finally
-            {
-                semaphore.Release();
-            }
-        }
-
-        class MessageState
-        {
-            public IModel Channel { get; set; }
-
-            public ulong DeliveryTag { get; set; }
         }
 
-        Task AcknowledgeMessage(IModel channel, ulong deliveryTag)
+        private async Task TryAcknowledgeMessage(ulong deliveryTag, string messageId)
```

Comment:
Also, I tend to prefer to leave the "Try*" naming scheme for methods that actually return a bool and have an out parameter, since that is the established pattern for those in the framework.